src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs

```diff
@@ -326,6 +330,34 @@ internal virtual NamedTypeSymbol LookupMetadataType(ref MetadataTypeName emitted
             }
 
 Done:
+            if (isTopLevel
+                && scope is not PENamespaceSymbol
+                && GeneratedNameParser.TryParseFileTypeName(
+                    emittedTypeName.UnmangledTypeName,
+                    out string? displayFileName,
+                    out int ordinal,
+                    out string? sourceName))
+            {
+                // also do a lookup for file types from source.
+                namespaceOrTypeMembers = scope.GetTypeMembers(sourceName);
+                foreach (var named in namespaceOrTypeMembers)
+                {
+                    if (named.AssociatedSyntaxTree is SyntaxTree tree
+                        && getDisplayName(tree) == displayFileName
+                        && named.DeclaringCompilation.GetSyntaxTreeOrdinal(tree) == ordinal
+                        && named.Arity == emittedTypeName.InferredArity)
+                    {
```

Comment:
The requested `emittedTypeName` may have `UseCLSCompliantNameArityEncoding` and `ForcedArity`. Feels like we need to deal with those.